VMC Test scenarios

marunachalam opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
*Test Scenarios for Deploy Update and Scale Deployments for VMC

Scale up and scale down deployments
Deployments with volume provisioning

Execution logs:
https://gist.github.com/marunachalam/c9b32f48c41a37bb79e621b356d19f71

marunachalam commented:
jtest wcp

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: FAILURE 
Stage before exit: null 

```

svcbot-qecnsdp commented:
Started WCP block pipeline...

svcbot-qecnsdp commented:
``` 
WCP build status: SUCCESS 
Stage before exit: finally 
Jenkins E2E Test Results: 

Ran 32 of 185 Specs in 6419.090 seconds
SUCCESS! -- 32 Passed | 0 Failed | 0 Pending | 153 Skipped
PASS

Ginkgo ran 1 suite in 1h47m26.841135764s
Test Suite Passed
make: Leaving directory `/home/worker/workspace/github-csi-wcp-CICD/Results/661/vsphere-csi-driver`

```

marunachalam commented:
jtest gc

svcbot-qecnsdp commented:
Started GC block pipeline...

svcbot-qecnsdp commented:
``` 
GC build status: FAILURE 
Stage before exit: replace-image-in-vc 

```

sashrith commented on tests/e2e/util.go:
```diff
@@ -2348,6 +2350,91 @@ func deleteFcdWithRetriesForSpecificErr(ctx context.Context, fcdID string, dsRef
 }
 
+// NewDeploymentwithVolume returns a deployment spec with the specified argument.
+func NewDeploymentwithVolume(deploymentName string, replicas int32, podLabels map[string]string, imageName, image string, pvclaims []*v1.PersistentVolumeClaim, strategyType appsv1.DeploymentStrategyType) *appsv1.Deployment {
```
See if `createDeployment()` is sufficient instead of the three methods we have added here

marunachalam replied:
yes I have utilised your `createDeployment()` method which is added recently

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,215 @@
+		var lables = make(map[string]string)
+		lables["app"] = "nginx"
+		deploymentName = "testdeployment"
+		replica = 3
```
If this is for block volume, then it will not work since block volume are RWO, same volume cannot be used by different pods(replicas).

marunachalam replied:
As discussed I have removed this testcase

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,215 @@
+		dep, err := client.AppsV1().Deployments(namespace).Get(ctx, deploymentName, metav1.GetOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		framework.Logf("replica count after scale down of deployments %v", dep.Status.Replicas)
+		gomega.Expect((dep.Status.Replicas) == maxReplica).To(gomega.BeTrue(), "Number of containers in the deployment should match with number of replicas")
```
We should also verify if the required number of replicas are running ready as well

marunachalam replied:
as discussed I removed this test cases

sashrith reviewed (changes requested):
this test needs a good amount of rework

svcbot-qecnsdp commented:
Started GC block pipeline...

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,180 @@
+	fpv "k8s.io/kubernetes/test/e2e/framework/pv"
+)
+
+var _ = ginkgo.Describe("[vmc-gc] [csi-guest] Deploy Update and Scale Deployments", func() {
```
Deploy, update - typo

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,180 @@
+		svcClient, svNamespace = getSvcClientAndNamespace()
+		setResourceQuota(svcClient, svNamespace, rqLimit)
+
+		sc, err := client.StorageV1().StorageClasses().Get(ctx, storageclassname, metav1.GetOptions{})
+		if err == nil && sc != nil {
+			gomega.Expect(client.StorageV1().StorageClasses().Delete(ctx, sc.Name, *metav1.NewDeleteOptions(0))).NotTo(gomega.HaveOccurred())
+		}
```
why is this required?

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,180 @@
+		defer func() {
+			err := fpv.DeletePersistentVolumeClaim(client, pvclaim.Name, namespace)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			pvclaim = nil
```
wait for cns volume to be deleted

sashrith commented on tests/e2e/util.go:
```diff
@@ -2348,6 +2347,26 @@ func deleteFcdWithRetriesForSpecificErr(ctx context.Context, fcdID string, dsRef
 }
 
+//updateDeploymentImage updates the image in the deployment with the given image
+func updateDeploymentImage(ns string, image string, deploymentName string, imageName string) {
```
Please rework on this method.

You can use `UpdateDeploymentWithRetries()` and `WaitForDeploymentCompleteAndCheckRolling()`

sashrith commented on tests/e2e/vmc_csi_deployments.go:
```diff
@@ -0,0 +1,180 @@
+
+		if pvclaim != nil {
+			err := fpv.DeletePersistentVolumeClaim(client, pvclaim.Name, namespace)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
```
wait for cns volume to be deleted
Also, delete deployment here, if it was not deleted in test case

marunachalam commented:
Unable to rebase this PR because of conflicts,
I have created another PR
https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1242
addressed all the review comments
